File: backend/app/schemas/progress.py
from datetime import date
from typing import Literal

from pydantic import BaseModel, ConfigDict

# ---------------------------------------------------------------------------
# Weight log
# ---------------------------------------------------------------------------

WeightSource = Literal["prompt", "manual", "baseline"]


class WeightLogEntryCreate(BaseModel):
    date: date
    weight_kg: float
    source: WeightSource = "manual"


class WeightLogEntryRead(BaseModel):
    date: date
    weight_kg: float
    source: str

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Archived goals
# ---------------------------------------------------------------------------


class ArchivedGoalCreate(BaseModel):
    """
    Matches the ArchivedGoalSummary type in the frontend.
    The id is a stable string derived from startDate_targetDate.
    """

    id: str
    start_date: date
    target_date: date
    start_weight_kg: float
    target_weight_kg: float
    end_date: date
    final_weight_kg: float | None = None
    weight_change_kg: float | None = None
    archived_at: date


class ArchivedGoalRead(ArchivedGoalCreate):
    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Goal snapshot fields (stored on User, updated via /users/me)
# ---------------------------------------------------------------------------


class GoalSnapshotUpdate(BaseModel):
    goal_start_date: date | None = None
    goal_start_weight_kg: float | None = None

File: frontend/types/progress.ts
/**
 * Progress feature types — mirrors the backend Pydantic schemas in
 * app/schemas/progress.py. Once `pnpm generate-client` is run these will
 * be in api/types.gen.ts and this file can be removed.
 */

export type WeightLogEntryRead = {
  date: string; // YYYY-MM-DD
  weight_kg: number;
  source: string;
};

export type ArchivedGoalRead = {
  id: string;
  start_date: string;
  target_date: string;
  start_weight_kg: number;
  target_weight_kg: number;
  end_date: string;
  final_weight_kg: number | null;
  weight_change_kg: number | null;
  archived_at: string;
};
